sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -88,6 +88,16 @@ abstract class BinaryArithmetic extends BinaryOperator {
 
   override def dataType: DataType = left.dataType
 
+  // Whether the dataType is promoted or not, we should only promote DecimalType once
+  var decimalPromoted: Boolean = false
```

Comment:
can you explain why this is needed? In addition, I'm not sure about adding a non-execution mutable state here.